### PR TITLE
Further RxSwift changes.

### PR DESCRIPTION
**Pull request by ashfurrow**

### CODE REVIEW

File: Kiosk/App/BidderDetailsRetrieval.swift [hunk 1]
```diff
@@ -39,6 +35,27 @@ extension UIViewController {
...
+    func emailPromptAlertController() -> UIAlertController {
+        let alertController = UIAlertController(title: "Send Bidder Details", message: "Enter your email address or phone number registered with Artsy and we will send your bidder number and PIN.", preferredStyle: .Alert)
+
+        let ok = UIAlertAction.Action("OK", style: .Default)
+        let action = CocoaAction { _ -> Observable<Void> in
+            let text = (alertController.textFields?.first)?.text ?? ""
+
+            return self
+                .retrieveBidderDetailsSignal(text)
+                .map(void)
```
Comment (orta) [inline Kiosk/App/BidderDetailsRetrieval.swift]: what's this map void do?

File: Kiosk/Bid Fulfillment/ConfirmYourBidArtsyLoginViewController.swift [hunk 2]
```diff
@@ -83,14 +99,19 @@ class ConfirmYourBidArtsyLoginViewController: UIViewController {
         }
     }
 
+    override func viewWillDisappear(animated: Bool) {
+        super.viewWillDisappear(animated)
+        _viewWillDisappear.onNext()
```
Comment (orta) [inline Kiosk/Bid Fulfillment/ConfirmYourBidArtsyLoginViewController.swift]: this seems like a pretty acceptable way to do these signals :+1:

File: Kiosk/App/BidderDetailsRetrieval.swift (hunk 1 above)
Comment (ashfurrow) [inline Kiosk/App/BidderDetailsRetrieval.swift]: Huh good question. It's purpose is to map Observables to send Void dos satisfy the type checker, but this observable is already Void. I'll remove it, thanks!

File: Kiosk/Bid Fulfillment/ConfirmYourBidArtsyLoginViewController.swift (hunk 2 above)
Comment (ashfurrow) [inline Kiosk/Bid Fulfillment/ConfirmYourBidArtsyLoginViewController.swift]: Yeah, I thought about abstracting it a bit, but I'm going to wait on https://github.com/ReactiveX/RxSwift/issues/181 to see if someone else fixes it for me :wink:

File: Kiosk/Help/HelpViewController.swift [hunk 3]
```diff
@@ -120,14 +127,16 @@ private extension HelpViewController {
...
         
-        hasBuyersPremiumSignal.subscribeNext { [weak self] in
-            let hasBuyersPremium = $0 as! Bool
-            if hasBuyersPremium {
-                self?.stackView.addSubview(self!.buyersPremiumButton, withTopMargin: "\(self!.inbetweenMargin)", sideMargin: "\(self!.sideMargin)")
-            } else {
-                self?.stackView.removeSubview(self!.buyersPremiumButton)
+        hasBuyersPremiumSignal
+            .subscribeNext { [weak self] in
+                let hasBuyersPremium = $0
```
Comment (orta) [inline Kiosk/Help/HelpViewController.swift]: ( nitpick, not blocker ) it feels weird seeing `$0` on multi line blocks, am I alone here?

Comment (ashfurrow) [inline Kiosk/Help/HelpViewController.swift]: Not alone, this was copied from the old version (see it above). I think I was in a hurry when I wrote it originally – I'll change it now :smile:

**Comment by ashfurrow** [comment]
Very close to app compiling – this PR is ready for review.

**Comment by orta** [comment]
Sounds like you'd really enjoy https://github.com/orta/You-Can-Do-It/

**Comment by ashfurrow** [comment]
Ha, swell idea!

**Comment by ashfurrow** [comment]
OK, so this compiles. I'm working on getting it to _work_, in the meantime @orta could you take a look?

File: Kiosk/Admin/AdminCardTestingViewController.swift [hunk 4]
```diff
@@ -21,27 +21,28 @@ class AdminCardTestingViewController: UIViewController {
...
+                    self.log("\(message)")
+                case .Error(let error):
+                    self.log("\n====Error====\n\(error)\nThe card reader may have become disconnected.\n\n")
+                    if self.cardHandler.card != nil {
+                        self.log("==\n\(self.cardHandler.card!)\n\n")
+                    }
+                case .Completed:
+                    if let card = self.cardHandler.card {
+                        let cardDetails = "Card: \(card.name) - \(card.last4) \n \(card.cardToken)"
+                        self.log(cardDetails)
```
Comment (orta) [inline Kiosk/Admin/AdminCardTestingViewController.swift]: I wonder if this thing can be a `guard`?

Comment (orta) [inline Kiosk/Admin/AdminCardTestingViewController.swift]: and I bet that `let string = thing` then logging was an Xcode/swift bug that may not exist anymore

File: Kiosk/Admin/ChooseAuctionViewController.swift [hunk 5]
```diff
@@ -15,9 +15,12 @@ class ChooseAuctionViewController: UIViewController {
...
 
-        XAppRequest(endpoint).filterSuccessfulStatusCodes().mapJSON().mapToObjectArray(Sale.self)
-            .subscribeNext({ [weak self] (activeSales) -> Void in
-                self!.auctions = activeSales as! [Sale]
+        XAppRequest(endpoint)
+            .filterSuccessfulStatusCodes()
+            .mapJSON()
+            .mapToObjectArray(Sale.self)
+            .subscribeNext { [weak self] (activeSales) -> Void in
+                self!.auctions = activeSales
```
Comment (orta) [inline Kiosk/Admin/ChooseAuctionViewController.swift]: could the `bang` code lose the `self!` via a strong ref inside here?

File: Kiosk/App/Models/SaleArtworkViewModel.swift [hunk 6]
```diff
@@ -119,4 +119,30 @@ extension SaleArtworkViewModel {
...
+            saleArtwork.rx_observe(NSNumber.self, "highestBidCents")
+        ]
+
+        return signals.combineLatest { numbers -> Int in
+            let bidCount = (numbers[0] ?? 0) as Int
+            let openingBid = numbers[1] as Int?
+            let highestBid = numbers[2] as Int?
+
+            return (bidCount > 0 ? highestBid : openingBid) ?? 0
+        }.map(centsToPresentableDollarsString)
```
Comment (orta) [inline Kiosk/App/Models/SaleArtworkViewModel.swift]: this is nice

File: Kiosk/App/Views/RegisterFlowView.swift [hunk 7]
```diff
@@ -25,15 +24,15 @@ class RegisterFlowView: ORStackView {
...
         return [
-            [SubViewParams(title: "Mobile", keypath: ["phoneNumber"])],
-            [SubViewParams(title: "Email", keypath: ["email"])],
-            [SubViewParams(title: "Postal/Zip", keypath: ["zipCode"])].filter { _ in self.appSetup.needsZipCode }, // TODO: may remove, in which case no need to flatten the array
-            [SubViewParams(title: "Credit Card", keypath: ["creditCardName", "creditCardType"])]
+            [SubViewParams(title: "Mobile", getters: [{ $0.phoneNumber.value }])],
+            [SubViewParams(title: "Email", getters: [{ $0.email.value }])],
+            [SubViewParams(title: "Postal/Zip", getters: [{ $0.zipCode.value }])].filter { _ in self.appSetup.needsZipCode }, // TODO: may remove, in which case no need to flatten the array
+            [SubViewParams(title: "Credit Card", getters: [{ $0.creditCardName.value }, { $0.creditCardType.value }])]
         ].flatMap {$0}
```
Comment (orta) [inline Kiosk/App/Views/RegisterFlowView.swift]: Doing my research

screen shot 2015-11-30 at 1 00 37 pm

Comment (orta) [inline Kiosk/App/Views/RegisterFlowView.swift]: I get this now, I also had to get ash over to talk about [this though](http://reactivex.io/documentation/operators/flatmap.html)

File: Kiosk/Bid Fulfillment/GenericFormValidationViewModel.swift [hunk 8]
```diff
@@ -1,22 +1,27 @@
...
-    init(isValidSignal: RACSignal, manualInvocationSignal: RACSignal, finishedSubject: RACSubject) {
-        command = RACCommand(enabled: isValidSignal) { _ -> RACSignal! in
-            return RACSignal.createSignal { (subscriber) -> RACDisposable! in
-                finishedSubject.sendCompleted()
-                subscriber.sendCompleted()
+    init(isValidSignal: Observable<Bool>, manualInvocationSignal: Observable<Void>, finishedSubject: PublishSubject<Void>) {
 
-                return nil
+        command = CocoaAction(enabledIf: isValidSignal) { _ in
+            return create { ovserver in
```
Comment (orta) [inline Kiosk/Bid Fulfillment/GenericFormValidationViewModel.swift]: `ovserver`

File: Kiosk/Bid Fulfillment/LoadingViewController.swift [hunk 9]
```diff
@@ -19,11 +19,16 @@ class LoadingViewController: UIViewController {
     @IBOutlet weak var backToAuctionButton: SecondaryActionButton!
     @IBOutlet weak var placeHigherBidButton: ActionButton!
 
+    private let _viewWillDisappear = PublishSubject<Void>()
+    var viewWillDisappear: Observable<Void> {
```
Comment (orta) [inline Kiosk/Bid Fulfillment/LoadingViewController.swift]: I get this pattern now, you're hiding the fact that it's actually a Observable that this class can trigger.

File: Kiosk/Bid Fulfillment/LoadingViewController.swift [hunk 10]
```diff
@@ -55,23 +60,47 @@ class LoadingViewController: UIViewController {
...
+            onDisposed: { [weak self] in
+                // Regardless of error or completion. hide the spinner.
+                self?.spinner.hidden = true
+            })
+            .addDisposableTo(rx_disposeBag)
+    }
+
+    override func viewWillDisappear(animated: Bool) {
+        super.viewWillDisappear(animated)
+        _viewWillDisappear.onNext()
```
Comment (orta) [inline Kiosk/Bid Fulfillment/LoadingViewController.swift]: does this just send a `nil` down the chain?

Comment (orta) [inline Kiosk/Bid Fulfillment/LoadingViewController.swift]: or are you calling the "private" method directly here?

File: Kiosk/Bid Fulfillment/PlaceBidViewController.swift [hunk 11]
```diff
@@ -101,7 +129,13 @@ class PlaceBidViewController: UIViewController {
                     let lotNumberLabel = smallSansSerifLabel()
                     lotNumberLabel.tag = LabelTags.LotNumber.rawValue
                     detailsStackView.addSubview(lotNumberLabel, withTopMargin: "10", sideMargin: "0")
-                    RAC(lotNumberLabel, "text") <~ saleArtwork.viewModel.lotNumberSignal.takeUntil(viewWillDisappearSignal())
+                    saleArtwork.viewModel
+                        .lotNumberSignal()
+                        .filterNil()
```
Comment (orta) [inline Kiosk/Bid Fulfillment/PlaceBidViewController.swift]: aww, what a nice little method.

File: Kiosk/Bid Fulfillment/RegisterViewController.swift [hunk 12]
```diff
@@ -26,28 +31,42 @@ class RegisterViewController: UIViewController {
...
         coordinator.storyboard = self.storyboard!
-        let registerIndexSignal = RACObserve(coordinator, "currentIndex").takeUntil(viewWillDisappearSignal())
-        let indexIsConfirmSignal = registerIndexSignal.map { return ($0 as! Int == RegistrationIndex.ConfirmVC.toInt()) }
-        
-        RAC(confirmButton, "hidden") <~ indexIsConfirmSignal.not()
-        RAC(flowView, "highlightedIndex") <~ registerIndexSignal
+        let registerIndexSignal = coordinator.currentIndex
+        let indexIsConfirmSignal = registerIndexSignal.map { return ($0 == RegistrationIndex.ConfirmVC.toInt()) }
+
+        indexIsConfirmSignal
```
Comment (orta) [inline Kiosk/Bid Fulfillment/RegisterViewController.swift]: `indexIsConfirmSignal` sounds off, `indexIsConfirmedSignal`?

**Comment by orta** [comment]
Cool, all my comments are nitpicks, you're welcome to merge.

File: Kiosk/Bid Fulfillment/PlaceBidViewController.swift (hunk 11 above)
Comment (ashfurrow) [inline Kiosk/Bid Fulfillment/PlaceBidViewController.swift]: It's my fav :tada:

File: Kiosk/Bid Fulfillment/LoadingViewController.swift (hunk 10 above)
Comment (ashfurrow) [inline Kiosk/Bid Fulfillment/LoadingViewController.swift]: It's actually sending `Void()`, which is just an empty tuple `()`, so Swift infers from the empty parameter list that that's what I want to send (only for `Void` parameter types).

File: Kiosk/Bid Fulfillment/LoadingViewController.swift (hunk 9 above)
Comment (ashfurrow) [inline Kiosk/Bid Fulfillment/LoadingViewController.swift]: Yeah exactly – only the `LoadingViewController` can mutate the observable.

File: Kiosk/Admin/ChooseAuctionViewController.swift (hunk 5 above)
Comment (ashfurrow) [inline Kiosk/Admin/ChooseAuctionViewController.swift]: Sure, good idea.

**Pull request merged**